facebookads/api.py

```diff
@@ -500,25 +500,26 @@ def execute(self):
         retry_indices = []
 
         for index, response in enumerate(responses):
-            if response:
-                body = response.get('body')
-                code = response.get('code')
-                headers = response.get('headers')
-
-                inner_fb_response = FacebookResponse(
-                    body=body,
-                    headers=headers,
-                    http_status=code,
-                    call=self._batch[index],
-                )
-
-                if inner_fb_response.is_success():
-                    if self._success_callbacks[index]:
-                        self._success_callbacks[index](inner_fb_response)
-                elif self._failure_callbacks[index]:
-                    self._failure_callbacks[index](inner_fb_response)
-            else:
+            if not response:
                 retry_indices.append(index)
+                continue
+
+            inner_fb_response = FacebookResponse(
+                body=response.get('body'),
+                headers=response.get('headers'),
+                http_status=response.get('code'),
+                call=self._batch[index],
+            )
+            if inner_fb_response.is_success() and self._success_callbacks[index]:
+                self._success_callbacks[index](inner_fb_response)
+            else:
+                # retry transient errors
+                if inner_fb_response.is_transient():
+                    retry_indices.append(index)
+
+                # execute failure callbacks on error, even if transient
+                if self._failure_callbacks[index]:
```

Comment:
Thanks for explaining! The scenario you described is definitely a common case.

However, from SDK point of view, we should be more careful.
Before this change, the function execute has three possible outcomes:
1. succeed: execute _success_callbacks if exist.
2. no response: add to the returned batch.
3. fail: execute _failure_callbacks

With this PR, we're adding one more case: transient failure, which has a total different behavior than any of the original one; it will be added to the batch, but it'll also trigger the failure_callback.

Transient failure is not the same as total fail, but there's no option for callers to distinguish what they want to do for total failure and transient failure.

So we have three options here:
1. categorize transient failure with one of the original outcome and align with the original outcome behavior, which will be, just adding to the retry batch and not call failure_callback.
2. categorize transient failure with one of the original outcome and change the original outcome behavior with the transient failure, which will be, call failure_callbacks even with no response result.
3. add a new outcome. In this case, it'll be adding another __transient_error_callbacks.

Option 3 gives any caller the most flexibility to do whatever they want.

How do you think?